pkg/body_extractor/grpc_payload.go

```diff
@@ -117,30 +138,152 @@ func fieldFromProtoMessage(msg []byte, tagIndex int) (string, error) {
+			lastBuilderMsg = builder.NewMessage(fmt.Sprintf("msg%s%d", fieldName, subQuery.Field)).AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, subQuery.Field),
+				builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(subQuery.Field)).SetRepeated())
+		} else if subQuery.DataType == Message {
+			lastBuilderMsg = builder.NewMessage(fmt.Sprintf("msg%s%d", fieldName, subQuery.Field)).AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, subQuery.Field),
+				builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(subQuery.Field)))
+		} else if subQuery.DataType == NestedArray {
+			lastBuilderMsg = builder.NewMessage(fmt.Sprintf("msg%s%d", fieldName, subQuery.Field)).AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, subQuery.Field),
+				builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(subQuery.Field)).SetRepeated())
+		}
+		fieldName = "_" + fieldName
+	}
+
+	if queries[0].DataType == Message {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(queries[0].Field)))
+	} else if queries[0].DataType == NestedArray {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(queries[0].Field)).SetRepeated())
+	} else if queries[0].DataType == String {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(queries[0].Field)))
+	} else if queries[0].DataType == StringArray {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(queries[0].Field)).SetRepeated())
+	}
 
-func buildPayloadGenericProto() (*desc.MessageDescriptor, error) {
-	if genericProtoCache != nil {
-		return genericProtoCache, nil
+	msgDescriptor, err := builderMsg.Build()
+	if err != nil {
+		return nil, err
 	}
+	return msgDescriptor, nil
+}
+
+func RunQuery(msg *dynamic.Message, query []Query) (interface{}, error) {
+	first := query[0]
+	rest := query[1:]
 
-	builderMsg := builder.NewMessage("message")
-	for i := 1; i < 100; i++ {
-		builderMsg.AddField(builder.NewField(fmt.Sprintf("field_%d", i),
-			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(i)))
+	var val interface{}
+	var err error
+	if first.Index != -1 {
+		val, err = msg.TryGetRepeatedFieldByNumber(first.Field, first.Index)
+	} else {
+		val, err = msg.TryGetFieldByNumber(first.Field)
 	}
-	desc, err := builderMsg.Build()
 	if err != nil {
 		return nil, err
 	}
-	genericProtoCache = desc
-	return genericProtoCache, nil
+
+	if len(rest) == 0 {
+		return val, nil
+	}
+
+	dm, ok := val.(*dynamic.Message)
+	if !ok {
```

Comment:
do we know in what cases this will not be ok?

Reply:
not sure about the cases, but considering `msg.TryGetFieldByNumber` and `TryGetRepeatedFieldByNumber` both return back `interface{}`, this is just a safety net.